### ISSUE TITLE
[BUG] Spline-based semiparametric estimator won't train under random initialization

### ISSUE THREAD

**Issue by JamesCarzon** [lee-group-cmu/lf2i#39]
Note: The following is Claude Code's analysis of the silent failure exhibited in the semiparametric spline model, as seen on `dev` at 16bc944507a83ea07ae57867d7b2ff1593965695.

# Analysis: `parametric-nn` + `cdf_model="ispline"` P-Value Estimation

## Context

The user is finding the ispline CDF variant of `ParametricCDFEstimator` to perform poorly compared to the sigmoid variant. This document explains what the code is doing and diagnoses several concrete bugs/design flaws — from a showstopper initialization bug to subtler architectural problems.

**Files involved:**
- `src/lf2i/calibration/parametric_cd.py` — `ISplineCDFModel`, `ISplineWeightNetwork`, `ParametricCDFEstimator`
- `src/lf2i/calibration/p_values.py` — `estimate_rejection_proba` dispatcher (lines 263–270)
- `src/lf2i/calibration/isplines.py` — the `Isplines` basis-function evaluator
- `src/lf2i/inference/lf2i.py` — how LF2I wires inputs into `estimate_rejection_proba` (lines 217–239)

---

### Step 1 — How LF2I calls in (lf2i.py:217–239)

For `algorithm="parametric-nn"`, LF2I skips the `augment_calibration_set` augmentation step that other algorithms use. Instead it:
1. Stacks `[test_statistics_calib | parameters_calib]` into a single matrix `inputs_for_calib`
2. Passes dummy zero-filled `rejection_indicators` (they're noted as unused)
3. Calls `estimate_rejection_proba`, which constructs and fits a `ParametricCDFEstimator`

### Step 2 — What ParametricCDFEstimator.fit() does (parametric_cd.py:453–581)

The model learns to estimate the conditional CDF `F(λ; θ) = P(T ≤ λ | θ)`, where `λ` is the test statistic and `θ` is the parameter of interest. For `cdf_model="ispline"`:

**Normalisation:** The empirical min and max of the training test statistics are stored as `lambda_min_` and `lambda_max_`. All `λ` values are linearly mapped to `[0, 1]`.

**CDF model — `ISplineCDFModel`:**  
A weighted sum of I-spline basis functions:
```
F̃(λ; θ) = w_lower(θ) + Σ_k w_k(θ) · I_k(λ_norm)
```
- `I_k` are fixed, pre-computed monotone basis functions on [0, 1] (`Isplines` from isplines.py)
- `n_basis = n_knots - 2 + order = 6 - 2 + 3 = 7` by default
- `w_k(θ) ≥ 0` (enforced by softplus) and `w_lower(θ) ∈ (0,1)` (enforced by sigmoid)
- Importantly, I_k(0) = 0 for all k, so `F̃(0; θ) = w_lower(θ)`
- And I_k(1) = 1 for all k, so `F̃(1; θ) = w_lower(θ) + Σ_k w_k(θ)` — **unconstrained and can exceed 1**

**CDF weight network — `ISplineWeightNetwork`:**  
A small feed-forward network `θ → (weights, w_lower)` via softplus/sigmoid.

**Training loss — Brier score (parametric_cd.py:18–27, 554–565):**  
The cross-product Brier score. For each mini-batch of size n, the model:
1. Evaluates the n × n matrix `cdf[i, j] = F̃(λ_j; β(θ_i))`
2. Forms the indicator matrix `indicators[i, j] = 1{λ_i ≤ λ_j}`
3. Minimises `mean_{i,j}[(F̃(λ_j; β(θ_i)) - 1{λ_i ≤ λ_j})²]`

The Brier score is the correct objective: the minimiser is `F̃(λ_j; θ_i) = P(T_i ≤ λ_j | θ_i)`, because `1{λ_i ≤ λ_j}` is an unbiased indicator for that probability (λ_i and λ_j come from independent i.i.d. draws).

### Step 3 — How p-values are computed (parametric_cd.py:585–641)

Given a grid of `(λ, θ)` pairs at test time:
1. Normalise `λ` using the stored `lambda_min_`, `lambda_max_`
2. Call `ISplineCDFModel.forward()` to get scalar CDF values
3. For `acceptance_region='left'`: `p-value = 1 − F̃(λ; θ)`  
   For `acceptance_region='right'`: `p-value = F̃(λ; θ)`

---

### Bug 1 — Gradient vanishing at initialisation (CRITICAL)

This is the dominant cause of poor performance. The CDF at the right boundary is:
```
F̃(1; θ) = sigmoid(out_lower) + Σ_k softplus(out_k)
```

At random initialisation (Kaiming uniform, near-zero network outputs):
- `sigmoid(0) ≈ 0.53`
- `softplus(0) ≈ 0.693` per weight, × 7 weights = 4.85

So `F̃(1; θ) ≈ 5.4` for every θ.

After `torch.clamp(cdf, 0, 1)`, **every single element of the n×n CDF matrix is 1.0 at initialisation**. The gradient of `clamp` is 0 outside [0,1], so **all 512² gradient terms are exactly zero**. The model literally cannot move from its starting state.

Measured empirically:
```
CDF at right boundary (should ≈ 1):  mean=5.21, all > 1
Fraction of CDF matrix clamped at init: 100%
```

The sigmoid CDF model does not have this problem — `sigmoid(·)` always outputs values in (0,1) by construction, so gradients always flow.

**Fix: replace softplus + clamp with softmax normalisation.** If weights are computed as `softmax(out_weights)` (i.e., they are non-negative and sum to 1), and w_lower is dropped or set to 0, then:
```
F̃(λ; θ) = Σ_k softmax_k(θ) · I_k(λ)
```
This is always in [0,1] without any clamping, because `Σ_k softmax_k = 1` and `I_k(λ) ∈ [0,1]`. Gradients always flow. The CDF is a proper mixture of CDFs — each I-spline is itself a valid CDF — so this is architecturally sound.

---

### Bug 2 — w_lower is initialised incorrectly (CRITICAL, related to Bug 1)

At initialisation, `w_lower = sigmoid(out_lower) ≈ 0.5`. This means the model says "the probability that a test statistic is below its empirical minimum is 50%". That is wildly wrong. The CDF at `lambda_min_` should be approximately 0.

Since sigmoid saturates near 0 only for large negative logits, the weight network has to produce strong negative outputs to learn that `w_lower ≈ 0` — which is slow and fights with the Kaiming-init scale.

**Fix (same as above): drop w_lower entirely.** Setting `w_lower = 0` is correct: the training data is normalised so lambda_min_ maps to 0, and we want `F̃(0; θ) = 0`. The `Isplines_total` class in isplines.py does allow a nonzero `w_lower` for extrapolation purposes, but for CDF estimation it should be 0.

---

### Bug 3 — Out-of-training-range test statistics are silently clipped (MODERATE)

At prediction time, test statistics from a new evaluation grid `θ` may fall outside `[lambda_min_, lambda_max_]` (the range seen during calibration). The normalisation:
```python
lam_norm = (X[:, 0] - self.lambda_min_) / (self.lambda_max_ - self.lambda_min_ + 1e-8)
```
can produce values outside [0,1]. Then `_eval_basis` clips them to [1e-6, 1-1e-6]:
```python
lam = np.clip(lambda_norm, 1e-6, 1.0 - 1e-6)   # parametric_cd.py:285
```

The effect: all test statistics below lambda_min_ are treated identically (mapped to the CDF at essentially 0), and all test statistics above lambda_max_ are treated identically (mapped to the CDF at essentially 1). This makes p-values **flat and uninformative** in the tails — the region where they matter most for testing near the critical level.

The `Isplines_total` class in isplines.py implements proper **linear extrapolation** outside the mesh, but `ISplineCDFModel._eval_basis` uses the simpler `Isplines` class and simply clips.

**Fix:** Use the full training data range plus some margin for normalisation, or use `Isplines_total` with extrapolation.

---

### Bug 4 — Brier loss is uniformly weighted over λ, but p-values care about the tails (MODERATE)

The cross-product Brier loss treats all evaluation levels `λ_j` equally. The loss at `λ_j = 0.5` (the median test stat) gets the same weight as the loss at `λ_j = 0.95` (the upper 5%). But for p-value computation at level α, we only need the CDF to be accurate near the α quantile of the test statistic.

The sigmoid model can use `WeightedPinballLoss` with a Gaussian weight centred at the target α level — this is much better for p-value accuracy. The ispline model cannot use pinball because it has no closed-form quantile, so it is locked to the uniform Brier loss.

**Fix:** Either (a) implement a Brier loss with non-uniform quadrature weights that emphasise the target tail, or (b) implement an approximate quantile function for the ispline by binary search, enabling pinball loss.

---

### Bug 5 — Dead diagonal in the Brier loss (MINOR)

The indicator matrix `indicators[i,j] = 1{λ_i ≤ λ_j}` has `indicators[i,i] = 1` always (since `λ_i ≤ λ_i`). So the diagonal always targets a CDF value of 1.0, regardless of where in the distribution `λ_i` falls. This biases the loss slightly toward higher CDF values. In large batches (n=512), the diagonal is 1/512 ≈ 0.2% of terms and is negligible, but it is a systematic bias.

---

## Summary table

| Issue | Severity | Root symptom |
|---|---|---|
| Gradient vanishing at init (softplus + clamp) | **CRITICAL** | Model cannot learn at all from cold start |
| w_lower initialised to ~0.5 (should be ~0) | **CRITICAL** | CDF floor is wrong; slow to correct |
| Out-of-range test stats silently clipped | **MODERATE** | Flat p-values in the tails |
| Brier loss uniformly weighted over λ | **MODERATE** | Poor p-value accuracy near target α |
| Dead diagonal in Brier loss | Minor | Slight systematic bias toward high CDF |

---

## Proposed fix for the architecture (summary)

Replace the current output layer and CDF evaluation with:

```python
# In ISplineWeightNetwork.forward():
out     = self.net(theta)                  # (n, n_basis)
weights = torch.softmax(out, dim=1)        # (n, n_basis), sums to 1, no clamping needed
return weights                             # drop w_lower entirely

# In ISplineCDFModel.forward():
B   = torch.FloatTensor(self._eval_basis(lambda_norm))  # (n, n_basis) in [0,1]
cdf = (B * weights).sum(dim=1)            # (n,) — always in [0,1], no clamp needed

# In ISplineCDFModel.forward_brier():
B   = torch.FloatTensor(self._eval_basis(lambda_norm))  # (n, n_basis)
cdf = weights @ B.T                       # (n, n_basis) @ (n_basis, n) = (n, n)
# always in [0,1], no clamp needed
```

This ensures:
- `F̃(0; θ) = 0` (basis functions are 0 at left boundary)
- `F̃(1; θ) = 1` (basis functions are 1 at right boundary, and softmax weights sum to 1)
- Gradients always flow (no clamp, softmax is everywhere differentiable)
- Valid CDF by construction at every training step

## Verification

After implementing the fix, run a synthetic test where the true conditional CDF is known (e.g., test statistic ~ χ²(θ) for θ ∈ [1, 10]):
1. Fit `ParametricCDFEstimator(cdf_model='ispline', ...)` and `(cdf_model='sigmoid', ...)`
2. Plot `F̃(λ; θ)` for several fixed θ values against the true χ² CDF — the ispline should track it closely
3. Check coverage: run `lf2i.diagnostics()` and verify the empirical coverage of 90% sets is near 90%
4. Check that training loss decreases smoothly from epoch 1 (the gradient-vanishing bug causes a flat loss curve early on — this should no longer occur)

**Comment by JamesCarzon** [comment]
This option is being deprecated in favor of other regression models.
